TST: Turn some tests with loops into parametrized tests.

DWesl opened the pull request:
I wanted to mark only some parts of the loops as xfail for another PR.  That part of the PR probably won't make it into numpy, but parametrized tests give better information on failure than tests with loops do, especially if the test segfaults during execution, so I'm submitting these here.

<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      http://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      http://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      http://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

DWesl committed:
TST: Turn some tests with loos into parametrized tests.

I wanted to mark only some parts of the loops as xfail for another PR.
That part of the PR probably won't make it into numpy, but I think
parametrized tests give better information on failure than tests with
loops do, so I'm submitting these here.

charris commented:
I think this could be made better looking using stacked parametrizations, the product/nested iteration, for instance, can be done
```
@pytest.mark.parametrize("one", [...])
@pytest.mark.parametrize("two", [...])
test_me(one, two):
    pass
```

DWesl committed:
TST: Change big parametrized loop to nested small parametrized loops.

Suggested by charris for clarity.

Since I'm not marking some tests xfail based on the values of all
parameters, this looks nicer.

DWesl commented:
> I think this could be made better looking using stacked parametrizations, the product/nested iteration, for instance, can be done
> 
> ```python
> @pytest.mark.parametrize("one", [...])
> @pytest.mark.parametrize("two", [...])
> def test_me(one, two):
>     pass
> ```

Done

charris commented:
Thanks @DWesl .
